### PR TITLE
evaluator: change math and time parma from interface to datum

### PR DESCRIPTION
**Pull request by zxylvlp**
Change math and time parma from interface to datum.

### CODE REVIEW

File: util/types/datum.go [hunk 1]
```diff
@@ -1087,6 +1129,43 @@ func (d *Datum) ToInt64() (int64, error) {
...
+func (d *Datum) ToFloat64() (float64, error) {
+	switch d.Kind() {
+	case KindInt64:
+		return float64(d.GetInt64()), nil
+	case KindUint64:
+		return float64(d.GetUint64()), nil
+	case KindFloat32:
+		return float64(d.GetFloat32()), nil
+	case KindFloat64:
+		return float64(d.GetFloat64()), nil
```
Comment (shenli) [inline util/types/datum.go]: return d.GetFloat64(), nil ?

File: evaluator/builtin.go [hunk 2]
```diff
@@ -104,16 +79,55 @@ var Funcs = map[string]Func{
...
+	"week":              {builtinWeek, 1, 2},
+	"weekday":           {builtinWeekDay, 1, 1},
+	"weekofyear":        {builtinWeekOfYear, 1, 1},
+	"year":              {builtinYear, 1, 1},
+	"yearweek":          {builtinYearWeek, 1, 2},
+	"extract":           {builtinExtract, 2, 2},
+	"date_arith":        {builtinDateArith, 3, 3},
+}
+
 func invArg(arg interface{}, s string) error {
```
Comment (shenli) [inline evaluator/builtin.go]: This function can be removed?

File: evaluator/builtin_test.go [hunk 3]
```diff
@@ -14,17 +14,90 @@
...
 
 import (
+	"fmt"
 	. "github.com/pingcap/check"
+	"github.com/pingcap/tidb/util/types"
+	"reflect"
 )
 
+// TblToDtbl is a util function for test.
+func TblToDtbl(i interface{}) []map[string][]types.Datum {
```
Comment (shenli) [inline evaluator/builtin_test.go]: TblToDtbl -> tblToDtbl?

File: evaluator/builtin_math.go [hunk 4]
```diff
@@ -28,54 +28,57 @@ import (
...
+	case types.KindNull:
+		return d, nil
+	case types.KindUint64:
+		return d, nil
+	case types.KindInt64:
 		// we don't need to handle error here, it must be success
-		v, _ := types.ToInt64(d)
-		if v >= 0 {
-			return x, nil
+		iv, _ := d.ToInt64()
```
Comment (coocood) [inline evaluator/builtin_math.go]: `d` is already int64, we don't need to call `ToInt64`.

**Comment by zxylvlp** [comment]
@coocood @shenli  PTAL

**Comment by coocood** [comment]
LGTM

File: evaluator/builtin_time.go [hunk 5]
```diff
@@ -425,33 +484,47 @@ func builtinDateArith(args []interface{}, ctx context.Context) (interface{}, err
...
-		return nil, ErrInvalidOperation.Gen("DateArith invalid args, need date but get %v", value)
+	if value.Kind() == types.KindNull {
+		d.SetNull()
+		return d, ErrInvalidOperation.Gen("DateArith invalid args, need date but get %v", value.GetValue())
 	}
-	result, ok := value.(mysql.Time)
-	if !ok {
-		return nil, ErrInvalidOperation.Gen("DateArith need time type, but got %T", value)
+	if value.Kind() != types.KindMysqlTime {
+		d.SetNull()
```
Comment (shenli) [inline evaluator/builtin_time.go]: d.SetNull()?

**Comment by shenli** [comment]
LGTM

**Pull request merged**